ecc/goldilocks/decaf.go

```diff
@@ -0,0 +1,135 @@
+
+// Neg calculates c=-a, where - is the inverse of the group operation.
+func (d Decaf) Neg(c, a *Elt) { c.p = a.p; c.p.cneg(1) }
+
+// Mul calculates c=n*a, where * is scalar multiplication on the group.
+func (d Decaf) Mul(c *Elt, n *Scalar, a *Elt) { d.c.ScalarMult(&c.p, n, &a.p) }
+
+// MulGen calculates c=n*g, where * is scalar multiplication on the group,
+// and g is the generator of the group.
+func (d Decaf) MulGen(c *Elt, n *Scalar) { d.c.ScalarBaseMult(&c.p, n) }
+
+// IsIdentity returns True if e is the identity of the group.
+func (e *Elt) IsIdentity() bool { return fp.IsZero(&e.p.x) }
+
+// IsEqual returns True if e=a, where = is an equivalence relation.
+func (e *Elt) IsEqual(a *Elt) bool {
+	l, r := &fp.Elt{}, &fp.Elt{}
+	fp.Mul(l, &e.p.x, &a.p.y)
+	fp.Mul(r, &a.p.x, &e.p.y)
+	fp.Sub(l, l, r)
+	return fp.IsZero(l)
+}
+
+// UnmarshalBinary if succeeds returns a Decaf element by decoding the first
+// DecafEncodingSize bytes of data.
+func (e *Elt) UnmarshalBinary(data []byte) error {
+	if len(data) < DecafEncodingSize {
+		return errInvalidDecoding
+	}
+
+	s := &fp.Elt{}
+	copy(s[:], data[:DecafEncodingSize])
+	isPositiveS := fp.Parity(s) == 0
+	p := fp.P()
+	isLessThanP := isLessThan(s[:], p[:])
+
+	s2, den, num := &fp.Elt{}, &fp.Elt{}, &fp.Elt{}
+	isr, altx := &fp.Elt{}, &fp.Elt{}
+	t0, t1 := &fp.Elt{}, &fp.Elt{}
+	x, y := &fp.Elt{}, &fp.Elt{}
+	one := fp.One()
+	fp.Sqr(s2, s)                       // s2  = s^2
+	fp.Sub(den, &one, s2)               // den = 1 + a*s^2
+	fp.Mul(t1, s2, &paramDTwist)        // t1 = d*s^2
+	fp.Add(t1, t1, t1)                  // t1 = 2*d*s^2
+	fp.Add(t1, t1, t1)                  // t1 = 4*d*s^2
+	fp.Sqr(t0, den)                     // num = (1 + a*s^2)^2
+	fp.Sub(num, t0, t1)                 // num = (1 + a*s^2)^2 - 4*d*s^2
+	fp.Mul(t0, t0, num)                 // t0 = num*den^2
+	isQR := fp.InvSqrt(isr, &one, t0)   // v = 1/sqrt(num*den^2)
+	fp.Mul(t1, den, isr)                // altx = isr*den
+	fp.Mul(t1, t1, s)                   // altx = s*isr*den
+	fp.Add(t1, t1, t1)                  // t1   = 2*s*isr*den
+	fp.Mul(altx, t1, &sqrtAMinusDTwist) // altx = 2*s*isr*den*sqrt(A-D)
+	isNegX := fp.Parity(altx)           // isNeg = sgn(altx)
+	fp.Neg(t0, isr)                     // t0 = -isr
+	fp.Cmov(isr, t0, uint(isNegX))      // if altx is negative then isr = -isr
+	fp.Sqr(x, isr)                      // x = isr^2
+	fp.Mul(x, x, den)                   // x = isr^2*den
+	fp.Mul(x, x, num)                   // x = isr^2*den*num
+	fp.Mul(x, x, s)                     // x = s*isr^2*den*num
+	fp.Add(x, x, x)                     // x = 2*s*isr^2*den*num
+	fp.Mul(y, isr, den)                 // y = isr*den
+	fp.Add(t0, &one, s2)                // t0 = 1 - a*s^2
+	fp.Mul(y, y, t0)                    // y = (1 - a*s^2)*isr*den
+	if !(isPositiveS && isLessThanP && isQR) {
+		return errInvalidDecoding
+	}
+	e.p.x, e.p.y, e.p.ta, e.p.tb, e.p.z = *x, *y, *x, *y, one
```

Comment:
This is almost constant time now.  The assignments above should be conditional moves on the expression in the if statement above.